### DISCUSSION TITLE
Filament v3 **BREAKING CHANGE** proposals

### DISCUSSION THREAD

**Discussion by pxlrbt** [filamentphp/filament#2639] [Ideas]
> **Note**
> This discussion should only contain feature requests that that needs breaking changes. Please open a separate discussion for general feature requests.

Thought we could collect some ideas for refactoring in V3.

Every idea that has been accepted will be added to the roadmap - https://github.com/orgs/filamentphp/projects/2/views/1

**Comment by ryangjchandler** [comment]
Remove deprecated `*Action` classes, they now run off of modifiers like `->button()`, etc.

**Comment by ManojKiranA** [comment]
Extract `datetime()` from `TextColumn` to Selparate class such as

1. `DateTimeColumn`
2. `DateColumn`
3. `TimeColumn`

**Reply by ryangjchandler** [reply]
@ManojKiranAppathurai What is the benefit of doing this?

**Reply by ManojKiranA** [reply]
IMO adding these classes will become more readable.

**Reply by ManojKiranA** [reply]
Something like this

```
DateTimeColumn::make('created_at')
->format('Ymd H:i A')
->localTimeZone(Auth::user()->time_zone),
```

**Reply by danharrin** [reply]
Honestly, we were trying to move away from classes and towards more powerful modifier methods. But you could always release a plugin with the "more readable" date time formatting options :)

**Reply by ManojKiranA** [reply]
@danharrin  you mean introduce more methods into traits ?

**Reply by danharrin** [reply]
Yeah, more methods on the base class instead of extra classes

**Comment by ManojKiranA** [comment]
Ability to add collapsible content in table something [like this](https://rappasoft.com/docs/laravel-livewire-tables/v2/columns/available-methods#content-collapsing)
![z1rWHzP](https://user-images.githubusercontent.com/30294553/174053645-75d7270c-136f-4fc5-923f-7548d37e9397.png)

Add ability to add sub text to column somthing like this

![image](https://user-images.githubusercontent.com/30294553/174054270-cd632763-9083-4265-a2d7-09ceb164ba74.png)

**Reply by pxlrbt** [reply]
The second one is already possible with:

`TextColumn::make('text')->getStateUsing(new HtmlString('<strong>First line</strong><br>Second line'))`

**Reply by ManojKiranA** [reply]
I Like the work aroud 😍

**Reply by margarizaldi** [reply]
> The second one is already possible with:
> 
> `TextColumn::make('text')->getStateUsing(new HtmlString('\<strong\>First line\</strong\><br>Second line'))`

What's the difference with `formatStateUsing()` ?

**Reply by patimio66** [reply]
"Collapsible content" in tables is the most lacking feature for me in terms of creating responsive (i.e. mobile) tables! Hopefully we'll se that soon 😍😍

**Reply by ralphjsmit** [reply]
For everyone participating, just saw this, you can now also do `TextColumn::make('text')->description(fn() => "second line")`.

**Reply by ManojKiranA** [reply]
This is added to v3 Roadmap.

**Comment by alvleont** [comment]
What about a Calendar View for resources?  Just in cases like this, should be better a Calendar view rather than a Table View. 

<img width="1322" alt="Captura de Pantalla 2022-06-22 a la(s) 10 55 58 a m" src="https://user-images.githubusercontent.com/47951932/175077189-fb8b7bf4-01ff-41be-98cc-2579c9a64099.png">

Useful for service appointments, meetings, and so on...

**Reply by saade** [reply]
You can use [filament-fullcalendar](https://github.com/saade/filament-fullcalendar) for this, let me know if you have any issues with it

**Reply by danharrin** [reply]
Yes I'd like to keep this as a plugin, we have enough core features to maintain already

**Reply by romaldyminaya** [reply]
Agree with @danharrin

**Comment by ManojKiranA** [comment]
Ability to do Option Group in `Select`

**Comment by xiaomlove** [comment]
Action is available anywhere on custom page without having to be registered in getActions()

**Comment by invaders-xx** [comment]
Add functionnality to modals to full screen resize

**Comment by ManojKiranA** [comment]
Streamline the translation files.

Recently I have seen lots and lots ot PR based on the translation. But the problem most of the translations are not uptodate. I would be nice to have separate repo for translations alone

**Reply by pxlrbt** [reply]
What would be the benefit of a separate repo?

**Reply by ManojKiranA** [reply]
So that It can be up to date

**Reply by pxlrbt** [reply]
How would it be more up-to-date if it was shipped via a separate repo than with multiple releases per week on the main repo?

**Reply by ManojKiranA** [reply]
@pxlrbt Something like this

**Reply by pxlrbt** [reply]
Something like what? 🤔

**Reply by ManojKiranA** [reply]
https://discord.com/channels/883083792112300104/956270111176679516/1016497906464858172

**Comment by ariaieboy** [comment]
tree structure for form fields

a common usage is the category tree.

**Reply by pxlrbt** [reply]
I guess it's better if this is a plugin instead.

**Comment by G3z** [comment]
Possibility to configure Repeater's header. This will make `->collapsible()` more usable as now the blocks are anonymous

**Reply by cheesegrits** [reply]
Don't know if this was added after you posted this, but I just added headers into repeaters with ...

https://filamentphp.com/docs/2.x/forms/fields#item-labels

**Comment by G3z** [comment]


**Reply by pxlrbt** [reply]
Don't think that will require a BC ;)

**Comment by ManojKiranA** [comment]
Out of the Box support for Multiple Guards

#2082
#2058

**Comment by NathanaelGT** [comment]
custom type for closure by using an invokable class

something like
```php
function (Get $get, Set $set) {
    $title = $get('title');

    $set('slug', Str::slug($title));
}
```

so we get code completion for parameters

**Comment by pxlrbt** [comment]
Make all `getLabel()`, `getHeading()`, `getTitle()` methods public for better access via plugins etc.

**Comment by zayedadel** [comment]
Add Vite integration its already Laravel's default bundler.
also tailwind changed their installation according to this adaption.
https://tailwindcss.com/docs/guides/laravel

it is becoming the standard but at the same time migrating back to mix shouldn't be a headache 
is that even achievable or should it be one or the other ?

**Reply by Z3d0X** [reply]
Vite support added in filament version `v2.14.0`
You can see the docs for it here: https://filamentphp.com/docs/2.x/admin/appearance#building-themes

**Comment by ariaieboy** [comment]
can we extract the assets of the filament and use them as npm packages? in version 3? 
since filament uses alpine v3 extracting the javascript assets as alpinejs plugin is not that hard.

**Reply by pxlrbt** [reply]
Why do you want to install them via npm when they come already through composer?

**Reply by ariaieboy** [reply]
it has a lot of benefits first we can use a lot of these useful alpinejs components in non-filament projects. 
and second of all, we can have control over loading these assets and customizing them or bundling them in any way that we want.

**Comment by ariaieboy** [comment]
we have 2 problems with filament right now in version 2.
the first one that is kinda big is the testing part of filament we need a lot more test helpers for filament. I think there is some WIP about this for version 2 but for version 3, we should cover a lot of edge cases about testing. 

the second problem with the filament is the documentation. it's sometimes ambiguous. for example, look at this example :
```
Alternatively, you can use the getOptionLabelUsing() method to transform the selected option's Eloquent model into a label. But please note, this is much less performant than using a virtual column:

use Filament\Forms\Components\Select;
use Illuminate\Database\Eloquent\Model;
 
Select::make('authorId')
    ->relationship('author', 'first_name')
    ->getOptionLabelFromRecordUsing(fn (Model $record) => "{$record->first_name} {$record->last_name}")
```
in the `Select` documentation it says you can use `getOptionLabelUsing()` to transform the selected option's Eloquent model into a label but in the example code it uses `getOptionLabelFromRecordUsing()` method :-?

I think we need a full review for the document section.

**Reply by pxlrbt** [reply]
1. Some testing helpers were already added. You can PR others if there are useful ones missing.
2. If you find errors like these, you can open a PR on the repo. Dan just restructured big parts of the docs. Errors like these can always happen. If you want to do a full review of the docs, go ahead.

**Reply by ariaieboy** [reply]
> 1. Some testing helpers were already added. You can PR others if there are useful ones missing.
> 2. If you find errors like these, you can open a PR on the repo. Dan just restructured big parts of the docs. Errors like these can always happen. If you want to do a full review of the docs, go ahead.

I am pretty new to the filament. and I just started my first project using filament. 
I would love to make some PRs but I need more time reading the project itself to understand the core of filament.

about the documentation, I think major releases are a perfect time to revisit the documentation and make some improvements to it because the maintainers gonna go throw the whole project to make changes and do refactor. refactoring and revisiting the document in this process is a lot easier.

a good document gonna help contributors to start helping the project more easily.

**Comment by oddvalue** [comment]
The table builder is currently inexorably linked the Eloquent builder and it's models. It would be really useful if there was a more primitive table class that just took rows to display that could then be extended to have the full features via Eloquent. That way it would be possible to use the table builder to put together simple tables for arrays without having to resort to something like https://github.com/calebporzio/sushi.

**Reply by danharrin** [reply]
You can use the table blade components to build custom tables if you want, they don't require you to use Livewire or Eloquent. But obviously your tables will not have functionality

**Reply by oddvalue** [reply]
Are we safe to assume that the blade components will maintain a consistent implementation then?

**Reply by pxlrbt** [reply]
I don't think any internal components are safe to stay consistent.

**Reply by oddvalue** [reply]
In that case it would be good to have something external and reliable to be able to create tables with.

**Reply by margarizaldi** [reply]
> In that case it would be good to have something external and reliable to be able to create tables with.

Even microsoft excel changes through time

**Comment by vv-ac** [comment]
1. Have a third level added to the navigation (similar to Nova, see also #2211).
2. Extend the layout by using the fullscreen, also similar to Nova, e. g. have the user menu moved to the very right. Related to this, see https://github.com/filamentphp/filament/discussions/774#discussioncomment-2778492.

Maybe that's not "breaking" enough?

Have a nice evening.

**Comment by namsudtho** [comment]
Horizontal menu design

**Comment by saade** [comment]
Nested resources?

**Comment by margarizaldi** [comment]
Optionally defer table filter queries until user click "Apply" instead of unnecessary instant query on each filter change. This is very useful especially for table that has a number of frequently used filters.

**Comment by margarizaldi** [comment]
`DaterangePicker` component with common presets like today, yesterday, last 7 days, etc.

**Reply by pxlrbt** [reply]
I think this would be better as a plugin

**Reply by margarizaldi** [reply]
For urgency I will agree with you, but when we already have DatePicker and DateTimePicker, why don't we have another popular and useful daterange (and datetimerange) picker as the native component in V3?

**Reply by pxlrbt** [reply]
Btw. this to totally unrelated to V3, as this in not a breaking change.

**Reply by margarizaldi** [reply]
ummm okay....

**Comment by margarizaldi** [comment]
Add possibility to create resource on-the-fly.

E.g.: In the `PostResource` create form, there is a select `Category` field, in this select options, the last option will be "create new", that will be the modal action to create new category resource.

I imagine the code will be like:
`Select::make('category')->relationship('category', 'name')->canCreate('Model', ...form schema...)`

**Reply by pxlrbt** [reply]
I think you are looking for `->createOptionForm()`:
https://filamentphp.com/docs/2.x/forms/fields#creating-new-records

**Reply by margarizaldi** [reply]
OMG how could I miss it? Thank you very much..

**Reply by blite** [reply]
@pxlrbt This currently doesn't work outside the admin right?

**Reply by danharrin** [reply]
It does?

**Comment by sbc640964** [comment]
I don't know where it was right to ask for it, I hope this is the place.

For now, you can create a custom page with custom built livewire content.

Is it possible to add the creation of a customized page with existing components from the package?

for example:
I need a settings page, my settings are stored in the settings table in the configuration of key and value, The value can be text, an array, or an image (etc.).

I am interested in sections and each setting as a separate custom form.

Today, my only option was to create a form and create the component and class for it, etc., etc.

I think it was easier to do it this way:
```php

public function getPageContent(){
   return [
      Section::make('One setting')->content([
        Form('one_setting')->schema([...])
      ]),
     ...
   ]
}

```

**Reply by pxlrbt** [reply]
> my only option was to create a form and create the component and class for it, etc., etc.

So you created a separate component for your form? Pages are Livewire components, and you can add your form to them.

> and each setting as a separate custom form.

A page can also have multiple forms. But you need to add the sections around them manually

**Reply by sbc640964** [reply]
I know, but I thought it would be simpler if it was from the php page created in Filament/Pages/MyPage, like a function that returns an array of components including tables, forms and more.

**Comment by sbc640964** [comment]
I think it would be useful to add a prediction option for the index column as cards/timeline/kanban board etc.

**Reply by sbc640964** [reply]
Basically give exposed access- headless - to the filtering and sorting functions and allow you to customize the list so that we can create plugins or to attach the list option to different views such as calendar / kanban / cards / and more...

**Reply by Z3d0X** [reply]
Hi @sbc640964  its already possible to provide a custom view for the table content https://filamentphp.com/tricks/use-custom-view-for-table-content

**Reply by sbc640964** [reply]
You're right, I've seen it before.
What is limited is that you are inside the white container and it is not suitable for any display

**Comment by dtraxx** [comment]
the possibility to store the selected columns in list view per user (if toggle-able)

**Reply by pxlrbt** [reply]
Isn't that already persisted in local storage? 🤔

**Reply by dtraxx** [reply]
in the DB like user preferences! i can't find it in local storage

**Reply by danharrin** [reply]
It's persisted in the session, you could use the DB driver for that

**Comment by margarizaldi** [comment]
### Defer Loading Table, Widget, ... everything

Every page should be loaded before/while performing table or widget queries. With an animated loading state shows until the data is fully loaded, what a cool admin panel.

**Comment by coleshirley** [comment]
make my [Filament Multi-Context](https://github.com/artificertech/filament-multi-context) package obsolete because the functionality is built in to the base package 🙏🏼

**Reply by danharrin** [reply]
Yeah I am down for exploring this

**Comment by khanalpride** [comment]
When using Relation Select with searchable, is it possible to auto show items (limit 10-15) by default without starting to type search query.

**Reply by danharrin** [reply]
Please only include breaking changes in this thread

This is already possible using `->preload()`

**Comment by IllesAprod** [comment]
Would be nice if every form component's schema can be added with ->schema function, it's hard to remember which component gets it from constructor parameter.

**Reply by danharrin** [reply]
That's already the case! You can omit the constructor parameter if you wish.

**Reply by danharrin** [reply]
If I've missed any, let me know

**Comment by SebastiaanKloos** [comment]
Add a settings page where we can add our custom settings from our packages to. Just like October CMS does

**Reply by danharrin** [reply]
I think this should be a plugin that other plugin authors can then hook into

**Comment by Z3d0X** [comment]
Make [this](https://github.com/filamentphp/filament/blob/4fc5666a3884b2f521b00647637d88a6f2c1c780/packages/forms/src/Components/Concerns/CanBeValidated.php#L247) method public

**Reply by danharrin** [reply]
I wouldnt class that as a breaking change, no-one would be overriding that. PR?

**Comment by zepfietje** [comment]
Configure icons using an `Icon` class instead of separate `icon()`, `iconColor()` and (potentially) `iconSize()` methods: https://github.com/filamentphp/filament/pull/3568.

**Comment by khanalpride** [comment]
I see there is option to have filters appear as column layout rather than a dropdown on the right side.
But is there a possibility that we can create our own filters layout, like creating form layout?

**Reply by danharrin** [reply]
What filters layout are you looking to achieve?

**Reply by khanalpride** [reply]
```
$table
            ->columns([
                Tables\Columns\TextColumn::make('title'),
                Tables\Columns\TextColumn::make('amount')->money('usd', true),
                Tables\Columns\TagsColumn::make('expenseTypes.title'),
                Tables\Columns\TextColumn::make('created_at')
                    ->dateTime('Y-m-d')
                    ->sortable(),
            ])
            ->filters([
                // grid layout the filters with our own sets of col width and layout
                Grid::make()
                    ->columns(3)
                    ->schema([
                        SelectFilter::make('Account')
                            ->column('account_id')
                            ->columnSpan(2)
                            ->relationship('account', 'title'),
                        MultiSelectFilter::make('Expense Types')
                            ->columnSpan(1)
                            ->relationship('expenseTypes', 'title'),
                    ]),
            ])
```

**Reply by margarizaldi** [reply]
> What filters layout are you looking to achieve?

Paired filters like min-max, date start-end, will be better to put together in single line.

**Reply by Z3d0X** [reply]
You can use `->form()` on `Filter` and use a grid in that to have filters like start & end date in the same line

```php
Filter::make('dates')
  ->form([
      Grid::make(2)
        ->schema([
            DatePicker::make('start_date')
                ->displayFormat('d-M-Y'),
  
            DatePicker::make('end_date')
                ->displayFormat('d-M-Y'),
        ]),
  ])
  ->query(function (Builder $query, array $data) {
      if (blank($data['start_date']) && blank($data['end_date'])) {
          return $query;
      }

      $query->when($data['start_date'], fn (Builder $query, $date) => $query->whereDate('created_at', '>=', $date))
          ->when($data['end_date'], fn (Builder $query, $date) => $query->whereDate('created_at', '<=', $date));

      return $query;
  }),
```

**Comment by MB116** [comment]
I would like to see in version 3 of Filament, built-in and advanced work with users, and permission management from the admin panel

**Reply by pxlrbt** [reply]
There is Shield and other plugins. I guess we will keep this out of core.

**Reply by danharrin** [reply]
Yes, the implementation of permissions is way too subjective for it to be in the core, policies are a good baseline

**Comment by mahfoudfx** [comment]
Extend the Select component to a multilevel tree checkbox.
![livewire-multi-level-checkbox.png](https://user-images.githubusercontent.com/31103481/187015497-810c4c01-557b-4001-91bf-b96626376065.png)

**Comment by deployHuman** [comment]
i think This class should be renamed 
to just FormInput or something, bc it feels weird and counter intuitive to look for a NumberInput in a TextInput class

https://github.com/filamentphp/filament/blob/2423ab235fc704d6f5939c5440acd69de828db66/packages/forms/src/Components/TextInput.php#L11

**Comment by ralphjsmit** [comment]
When creating custom forms and table components, they are created in the `resources/views/forms` and `resources/views/tables` folder. I personally view the `resources/views` folder really as a "top-level/high-overview" folder, so it feels a bit messy to have separate `forms` and `tables` folders. I'd prefer to clean it up and have them all inside the `resources/views/filament` folder. Thanks!

**Reply by ariaieboy** [reply]
i think we should move filament resource files to the folder that contain the pages and relation managers and etc.
the current folder structure is like this:
```
Filament/Resources
│   ExampleResource.php 
│
└───ExampleResource
│   │
│   └───Pages
│       │   EditExample.php
│       │   ...
```
the tree size get doubled because we have 1 directory for each resource and then the resource itself is in the root of `Filament/Resources` folder.
i think that the default structure must be sth like this:
```
Filament/Resources
│
└───ExampleResource
│   │ 
│   │ ExampleResource.php 
│   │ 
│   └───Pages
│       │   EditExample.php
│       │   ...
```

**Comment by fira03079508** [comment]
Per column filters. I saw that per column search was now available but you don't have an option to filter a specific column for a dange range or something like that

**Reply by margarizaldi** [reply]
@fira03079508 what do you mean? Filters mostly work per column already...

**Reply by fira03079508** [reply]
@margarizaldi I'm saying a filter button for a specific column. Example : You want to get rid of global filters and use only per column search/filter. You can do per column search at the moment but you can't do a filter for a specific column.
![image](https://user-images.githubusercontent.com/1340932/187533346-1cd57cca-6f5b-4564-9f6a-9099cde433bf.png)

![image](https://user-images.githubusercontent.com/1340932/187531283-b40d8980-996a-4b4b-88ea-d68dd16cb3c5.png)

**Reply by pxlrbt** [reply]
> You can do per column search at the moment but you can't do a filter for a specific column.

Well, you can. But yes, we don't have a "per column" UI 😅

**Reply by fira03079508** [reply]
Yeah that would be huge for our team.  (I've heard that it wouldn't be something that's "pluginable")

**Reply by margarizaldi** [reply]
@fira03079508 Well, I'm still thinking that most filters work per column, naturally, by default...

Yes we have global (app) search, table search & column search... But I don't think we've ever had kind of global filter... Really can't imagine how it works and what's the purpose of that "global filter"...

I probably just didn't catch your idea, therefore I'd like ask you what's the limitation of current filter ability? and let us know that this is not only about the matter of "filter placement"...

**Reply by fira03079508** [reply]
@margarizaldi as pxlrbt said. It's mostly a UI change. allowing filter to be associated to the column header UI wise. (a filer Icon for any column you want to apply a filter on, not just 1 filter button at the top to filter all the filterable columns)

**Reply by margarizaldi** [reply]
@fira03079508 to be honest, I have never met that kind of UI used in any apps I've ever used. None of e-commerce, crm, accounting or erp apps, mobile or desktop version, implement that design. Except spreadsheet. Correct me if I'm wrong.

In this case, having filter icons above each column is wasting space and make it looks dirty, while we can group all those filters into a single place.

I agree that we have an absolute freedom to dream our very ideal UI, but as a developer, I (personally) think it is always better to follow common best practices to make our users get familiar and comfort with our app as soon as possible.

Final words, just let the filament team to decide whether to accommodate this things or not. Who knows...

_Cheers..._

**Reply by fira03079508** [reply]
@margarizaldi It's used in a very common library : https://www.primefaces.org/primevue/datatable/filter

Also It's not an exact solution i'm asking for. all I wanted is to be able to have specific way to filter or search any column directly on the column.

I agree that separating the filters is a good thing. My team leaders really wants specific column search/filtering though ....

Here is an example of what it looks like at the moment on our side (we're looking to swap to Filament, this is with Yajra/datatable and YADCF 
![image](https://user-images.githubusercontent.com/1340932/187738349-6cb5a21d-b15d-42e6-b852-ab79a7548686.png)

**Comment by mahfoudfx** [comment]
As V3 is a Major Release, why not a UI lifting using [Flowbite](https://flowbite.com/) which compete directly with the official [TaiwlindCSS UI](https://tailwindui.com/).
You can have a look at the [Dashboard demo](https://flowbite.com/application-ui/preview/). As today, Razor UI has 100+ UI components and the last update is on 2021/11/23, where Flowbite has 450+ UI components and the last update is 2022/08/09.

![image](https://user-images.githubusercontent.com/31103481/187624155-665dbcd2-29f4-4c10-8b8d-7593cfd4c140.png)

**Reply by danharrin** [reply]
I doubt their license even allows it. We use Razor UI because Lars is a friend of the project and Filament has special permission to use the styles. We don't need to rely on any Razor UI updates, we can do those ourselves. We already have a customized version of Razor anyway.

**Reply by zepfietje** [reply]
And there's some UI tweaks on our radar already. So if you have any specific suggestions of things you'd like to see changed, let us know!

**Reply by mahfoudfx** [reply]
@danharrin it is under [MIT License](https://flowbite.com/docs/getting-started/license/) so it should be free to use. Is made this suggestion because I found that they have a large library which will allows the Filament team to free some UI efforts and focus more on core features rather than reinventing some wheels. Obviously, special permission with Razor UI it's also an advantage.
@zepfietje I will not hesitate to share my suggestions ;) .

**Reply by zepfietje** [reply]
Only a few Flowbite components are free.

**Reply by mahfoudfx** [reply]
> And there's some UI tweaks on our radar already. So if you have any specific suggestions of things you'd like to see changed, let us know!

@zepfietje By chance, do the UI tweaks on radar include animation when loading resources? Livewire already handle out of the box [Loading ](https://laravel-livewire.com/docs/2.x/loading-states) and [Offline](https://laravel-livewire.com/docs/2.x/offline-state) states.
Missing of animation gives the feeling of slowness when loading resources.

**Reply by zepfietje** [reply]
No. If there's anything specifically, feel free to discuss it on Discord or create a discussion/PR here on GitHub. Let's not bloat this thread with unrelated discussions.

**Comment by lloricode** [comment]
add loading on chart, or even queue for it

**Reply by zepfietje** [reply]
How exactly do you mean? And how is it a breaking change?

**Reply by lloricode** [reply]
we have a insight that have accessing a huge amount of data in database, so it takes time to load a whole site,

i just experience it in laravel nova that it load the site first then will add loading in widget part,

**Reply by zepfietje** [reply]
That's not something easy to implement with the nature of Filament and Livewire. Also, it's off-topic for this discussion. Feel free to open a separate discussion to explore ideas regarding this.

**Reply by romaldyminaya** [reply]
I'll try to replicate that behavior to see how difficult it would be to implement, because that adds a lot of freshness to the UX

**Comment by lloricode** [comment]


**Reply by zepfietje** [reply]
This thread is about changes that are breaking and thus should be added in v3. Let's keep the discussion on topic. You can create separate discussions or pull requests for new features that are non-breaking.

**Comment by ManojKiranA** [comment]
Ability to acces the `$loop` variable in Column or maybe introduce a new `IterationColumn` or someting.

https://discord.com/channels/883083792112300104/883083792653381695/978644639328391238

**Reply by danharrin** [reply]
Not a breaking change

**Comment by RX-77IT** [comment]
If a Model uses SoftDeletes, do you think it would be possible to make a trash bin button appear on top of the resource table? Pressing this button would show only the soft deleted models with the possibility to restore them.

**Reply by danharrin** [reply]
We already have a TrashedFilter which you can use to do that

https://filamentphp.com/docs/2.x/admin/resources/getting-started#handling-soft-deletes

Also not a breaking change

**Reply by RX-77IT** [reply]
Sorry, I totally missed that the handling of soft deletes is using filters. Thank you.

**Comment by kamilst96** [comment]
Maybe a table selector? Select and MultiSelect provide limited information, which can be problematic when searching through a bigger pile of records.

https://user-images.githubusercontent.com/37311362/189227675-79ee7716-d868-473a-99ae-9d9069208f5c.mp4

The workflow could be:
- In a form: click TableSelect > pick record from table (submit) > field is filled with selected record id.
- In a RM: click Attach / Associate > displays table with a form if added to schema > submit

**Reply by Z3d0X** [reply]
This can be a plugin

**Reply by ralphjsmit** [reply]
Will be launched as plugin! 👀

**Comment by MarJose123** [comment]
for the `Actions\ActionGroup::make` make the design as a group button as dropdown design.
